crates/data_components/src/s3_vectors/put_vectors_sink.rs

```diff
@@ -0,0 +1,428 @@
+            .ok_or_else(|| Error::ColumnTypeMismatch {
+                name: format!("data[{row}]"),
+                expected: "Float32Array".to_string(),
+            })?
+            .values()
+            .to_vec();
+
+        if vector.iter().any(|&x| x.is_nan() || x.is_infinite()) {
+            continue;
+        }
+
+        if vector.iter().all(|&x| x == 0.0) {
+            continue;
+        }
+
+        let mut metadata = HashMap::new();
+
+        for (index, name, data_type) in &fields {
+            let col = record_batch.column(*index);
+            let value = metadata_from_row(row, data_type, col);
+            metadata.insert((*name).to_string(), value);
+        }
+
+        let metadata = if metadata.is_empty() {
+            None
+        } else {
+            Some(Document::Object(metadata))
+        };
+
+        let put_input_vector = PutInputVector::builder()
+            .key(key)
+            .data(VectorData::Float32(vector))
+            .set_metadata(metadata)
+            .build()
+            .context(BuildInputSnafu)?;
+
+        put_input_vectors.push(put_input_vector);
+    }
+
+    Ok(put_input_vectors)
+}
+
+#[allow(clippy::unwrap_used)]
+fn metadata_from_row(row: usize, data_type: &DataType, col: &Arc<dyn Array + 'static>) -> Document {
+    match data_type {
+        DataType::Utf8 => {
+            let arr = col
+                .as_any()
+                .downcast_ref::<arrow::array::StringArray>()
+                .unwrap(); // UNWRAP: we checked the datatype
+            Document::String(arr.value(row).to_string())
+        }
+        DataType::Int64 => {
+            let arr = col
+                .as_any()
+                .downcast_ref::<arrow::array::Int64Array>()
+                .unwrap(); // UNWRAP: we checked the datatype
+            Document::Number(Number::NegInt(arr.value(row)))
+        }
+        DataType::UInt64 => {
+            let arr = col
+                .as_any()
+                .downcast_ref::<arrow::array::UInt64Array>()
+                .unwrap(); // UNWRAP: we checked the datatype
+            Document::Number(Number::PosInt(arr.value(row)))
+        }
+        DataType::Float64 => {
+            let arr = col
+                .as_any()
+                .downcast_ref::<arrow::array::Float64Array>()
+                .unwrap(); // UNWRAP: we checked the datatype
+            Document::Number(Number::Float(arr.value(row)))
+        }
+        DataType::Boolean => {
+            let arr = col
+                .as_any()
+                .downcast_ref::<arrow::array::BooleanArray>()
+                .unwrap(); // UNWRAP: we checked the datatype
+            Document::Bool(arr.value(row))
+        }
+        _ => unimplemented!(),
+    }
+}
+
+impl From<Error> for DataFusionError {
+    fn from(value: Error) -> Self {
+        DataFusionError::Execution(value.to_string())
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
```

Comment:
```suggestion
#[allow(clippy::expect_used)]
```

We allow expect in tests, but not unwrap